Packages/nadena.dev.modular-avatar/Editor/OptimizationPasses/GCGameObjectsPass.cs

```diff
@@ -23,11 +31,132 @@ internal GCGameObjectsPass(BuildContext context, GameObject root)
 
         internal void OnPreprocessAvatar()
         {
+            CollectReferences();
+            Cleanup();
             MarkAll();
             Sweep();
+            Debug.Log($"Cleanup complete, removed {_removedObjects} objects and {_removedComponents} components.");
         }
 
+        private void CollectReferences()
+        {
+            LoopAllComponents((obj, component) =>
+            {
+                switch (component)
+                {
+                    case SkinnedMeshRenderer mesh:
+                        CollectBonesFromSkinnedMeshRenderer(mesh);
+                        break;
+
+                    case VRCPhysBone pb:
+                        _physBones.Add(pb);
+                        break;
+                    case VRCPhysBoneCollider collider:
+                        _physBoneColliders.Add(collider);
+                        break;
+                }
+            });
+        }
+
+        private void CollectBonesFromSkinnedMeshRenderer(SkinnedMeshRenderer renderer)
+        {
+            HashSet<int> usedBonesIndices = new HashSet<int>();
+            foreach (var boneWeight in renderer.sharedMesh.boneWeights)
+            {
+                // technically should add only if weight is > 0, but not worth the cost and does not have negative effects
+                usedBonesIndices.Add(boneWeight.boneIndex0);
+                usedBonesIndices.Add(boneWeight.boneIndex1);
+                usedBonesIndices.Add(boneWeight.boneIndex2);
+                usedBonesIndices.Add(boneWeight.boneIndex3);
+            }
```

Comment:
It's probably better to use https://docs.unity3d.com/ScriptReference/Mesh.GetAllBoneWeights.html - this not only deals with meshes with >4 weights (though I'm not sure if VRC support that anyway), but also results in cleaner code as you don't need to have this 4x add thing.